openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
                     block_type=block_type, library_type=ref.type,
                 )
             )
-    lib_bundle = LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME)
-    # Total number of blocks should not exceed the maximum allowed
-    total_blocks = len(lib_bundle.get_top_level_usages())
-    if total_blocks + 1 > settings.MAX_BLOCKS_PER_CONTENT_LIBRARY:
+
+    # If adding a component would take us over our max, return an error.
+    component_count = publishing_api.get_all_drafts(ref.learning_package.id).count()
+    if component_count + 1 > settings.MAX_BLOCKS_PER_CONTENT_LIBRARY:
         raise BlockLimitReachedError(
-            _("Library cannot have more than {} XBlocks").format(settings.MAX_BLOCKS_PER_CONTENT_LIBRARY)
+            _("Library cannot have more than {} Components").format(
+                settings.MAX_BLOCKS_PER_CONTENT_LIBRARY
+            )
         )
+
     # Make sure the proposed ID will be valid:
-    validate_unicode_slug(definition_id)
+    validate_unicode_slug(block_id)
     # Ensure the XBlock type is valid and installed:
     XBlock.load_class(block_type)  # Will raise an exception if invalid
     # Make sure the new ID is not taken already:
-    new_usage_id = definition_id  # Since this is a top level XBlock, usage_id == definition_id
     usage_key = LibraryUsageLocatorV2(
         lib_key=library_key,
         block_type=block_type,
-        usage_id=new_usage_id,
+        usage_id=block_id,
     )
-    library_context = get_learning_context_impl(usage_key)
-    if library_context.definition_for_usage(usage_key) is not None:
-        raise LibraryBlockAlreadyExists(f"An XBlock with ID '{new_usage_id}' already exists")
-
-    new_definition_xml = f'<{block_type}/>'  # xss-lint: disable=python-wrap-html
-    path = f"{block_type}/{definition_id}/definition.xml"
-    # Write the new XML/OLX file into the library bundle's draft
-    draft = get_or_create_bundle_draft(ref.bundle_uuid, DRAFT_NAME)
-    write_draft_file(draft.uuid, path, new_definition_xml.encode('utf-8'))
-    # Clear the bundle cache so everyone sees the new block immediately:
-    BundleCache(ref.bundle_uuid, draft_name=DRAFT_NAME).clear()
+
+    if component_already_exists(usage_key):
+        raise LibraryBlockAlreadyExists(f"An XBlock with ID '{usage_key}' already exists")
+
+    _create_component_for_block(ref, usage_key)
+
     # Now return the metadata about the new block:
     LIBRARY_BLOCK_CREATED.send_event(
         library_block=LibraryBlockData(
             library_key=ref.library_key,
             usage_key=usage_key
         )
     )
+
     return get_library_block(usage_key)
 
 
-def delete_library_block(usage_key, remove_from_parent=True):
+def component_already_exists(usage_key: UsageKeyV2) -> bool:
```

Comment:
If this is a public API, it's weird to call it "already" exists, as it may be used in other cases besides creating a component.

This is also another good candidate for moving from the libraries API to the XBlock API. No library-specific code at the moment.

```suggestion
def component_exists(usage_key: UsageKeyV2) -> bool:
```

Reply:
Done

Reply:
Rather than adding this to the xblock api, I'm going to make it private, so that we can move it later without breaking the public API contract.